File: backend/src/projects/views/schemas/conflicts.py
from uuid import UUID

from pydantic import BaseModel

from src.projects.projects_db.schemas.weekday import WeekDay
from src.projects.services.schemas.conflicts import ConflictResult


class ConflictsResponse(BaseModel):
    conflicts: list[ConflictResult]
    count: int


class UpdateConflictTagRequest(BaseModel):
    tags: list[str]


class UpdateManyConflictTagItem(BaseModel):
    conflict_id: UUID
    tags: list[str]


class UpdateManyConflictTagsRequest(BaseModel):
    updates: list[UpdateManyConflictTagItem]


class UpdateManyConflictTagsResponse(BaseModel):
    conflicts: list[ConflictResult]


class ConflictPreviewRequest(BaseModel):
    original_block_id: UUID
    weekday: WeekDay
    start_time: int
    duration: int
    teacher_ids: list[UUID]
    room_ids: list[UUID]
    class_ids: list[UUID]


class ConflictPreviewResponse(BaseModel):
    solved: list[ConflictResult]
    new: list[ConflictResult]

File: frontend/src/types/project/conflicts.ts
// -- Base ----------------------------------------------------------------
export interface ConflictRecord {
  id: string;
  event_ids: string[];
  event_names: string[];
  day: string;
  time: number;
  turma: string[];
  block_ids: string[];
  degrees: string[];
  subjects: string[];
  conflict_reasons: string[];
  tags: string[];
}

// -- Tag constants -------------------------------------------------------
export const CONFLICT_TAGS = {
  PRE_EXISTING: "pre-existing",
  IGNORED: "ignored",
} as const;

// -- List payload --------------------------------------------------------
export interface ConflictsListPayload {
  conflicts: ConflictRecord[];
  count: number;
}

// -- Many tag update -----------------------------------------------------
export interface UpdateManyConflictTagItem {
  conflict_id: string;
  tags: string[];
}

export interface UpdateManyConflictTagsRequest {
  updates: UpdateManyConflictTagItem[];
}

export interface UpdateManyConflictTagsResponse {
  conflicts: ConflictRecord[];
}

// -- Preview -------------------------------------------------------------
export interface ConflictPreviewRequest {
  original_block_id: string;
  weekday: string;
  start_time: number;
  duration: number;
  teacher_ids: string[];
  room_ids: string[];
  class_ids: string[];
}

export interface ConflictPreviewResponse {
  solved: ConflictRecord[];
  new: ConflictRecord[];
}
